File: src/org/zkoss/jquery4j/jquery/Calculator.java
/* Calculator.java

	Purpose:
		
	Description:
		
	History:
	Apr 16, 2010	, Created by peterkuo

Copyright (C) 2006 Potix Corporation. All Rights Reserved.

{{IS_RIGHT
	This program is distributed under MIT in the hope that
	it will be useful, but WITHOUT ANY WARRANTY.
}}IS_RIGHT
 */

package org.zkoss.jquery4j.jquery;

import org.zkoss.zk.ui.event.Events;
import org.zkoss.zk.ui.event.InputEvent;
import org.zkoss.zul.impl.XulElement;

public class Calculator extends XulElement {
	
	double _value = 0;
		
	static {
		addClientEvent(Calculator.class, Events.ON_CHANGE, CE_IMPORTANT|CE_REPEAT_IGNORE);		
	}

	public Calculator(){}
	
	public Calculator(double value){
		this();
		setValue(value);
	}
	
	public double getValue(){
		return _value;
	}
	
	public void setValue(double value){
		if(_value!=value){
			_value = value;
			smartUpdate("value",value);
		}
	}	

	@Override
	protected void renderProperties(org.zkoss.zk.ui.sys.ContentRenderer renderer)
	throws java.io.IOException {
		super.renderProperties(renderer);	
		if(_value != 0){
			render(renderer, "value", Double.toString(_value));
		}
	}

	@Override
	public void service(org.zkoss.zk.au.AuRequest request, boolean everError) {		
		final String cmd = request.getCommand();
		if (cmd.equals(Events.ON_CHANGE)) {
			InputEvent evt = InputEvent.getInputEvent(request);
			
			final String value = evt.getValue();
			try {
				final double newval = Double.parseDouble(value);
				if(newval == _value){
					return; //not post event
				}					
				_value = newval;
			} catch (NumberFormatException ex) {
				throw new InternalError(value);
			}

			Events.postEvent(evt);
		} else {			
			super.service(request, everError);
		}		
	}
}

File: src/archive/web/js/jquery4j/jquery/Calculator.js
/* Calculator.js

{{IS_NOTE
	Purpose:

	Description:

	History:
	Apr 16, 2010	, Created by peterkuo
}}IS_NOTE

Copyright (C) 2008 Potix Corporation. All Rights Reserved.

{{IS_RIGHT
	This program is distributed under MIT in the hope that
	it will be useful, but WITHOUT ANY WARRANTY.
}}IS_RIGHT
*/

//because jquery ui rewrite the implementatoin of show of jquery.
//for other jquery widget not belong to jquery ui, shoule mark include jquery ui script in zk.wpd
jquery4j.jquery.Calculator = zk.$extends(zul.Widget,{
	
	_value: null,
	
	$define: {
		value: null
	},
	
	bind_: function () {
		this.$supers('bind_', arguments);
		var content = this.$n('cnt');		
		var wgt= this;		
		$(content).calculator({		
			onClose: function(value, inst) {
 				wgt.fire('onChange', {value: value});
    		}
		});	
	},
		
	redraw: function (out) {
		out.push('<div',this.domAttrs_(),'>');		
		out.push('<input id="',this.uuid,'-cnt" ','type="text"' );
		if(this._value){
			out.push(' value="', this._value ,'"')
		}			
		out.push('/></div>');
	}		
});